---
title: "Analysing cfRNA-seq Data"
author: 'Roman E. Reggiardo'
date: '09/05/2019'
output: html_notebook
---

All RNA-seq reads were processed with trimmed with *Trimmomatic*, quality controlled with *FastQC*, and psuedoaligned/aligned with *Salmon* and *STAR*, respectively.
```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 8, collapse = T)
knitr::opts_chunk$set(message = F)
```

## Load in:
```{r echo=FALSE}
# Run this cell to load in all packages used in the notebook

install.packages(c(
  'ggsci',
  'ggpubr',
  'ggsignif',
  'ggthemes', 
  'ggforce',
  'matrixStats',
  'ggrepel',
  'pheatmap',
  'viridis',
  'VennDiagram',
  'pheatmap'))


suppressPackageStartupMessages({
  library(tidyverse)
  library(ggsci)
  library(ggpubr)
  library(ggsignif)
  library(ggthemes) 
  #library(ggrepel)
  library(ggforce)
  #library(tximport)
  #library(tximeta)
  #library(extrafont)
  #library(msigdbr)
  #library(ggcorrplot)
  library(matrixStats)
  library(ggrepel)
  library(pheatmap)
  library(viridis)
  library(VennDiagram)
  #library(fgsea)
  #library(uwot)
  #library(NMF)
  #library(gt)
  #library(ggdendro)
})

#suppressMessages(loadfonts())

theme_set(theme_classic() + 
            theme(panel.background = element_rect(fill = 'white'),
                axis.text = element_text(family = 'Helvetica', size = 12),
                axis.title = element_text(family = 'Helvetica', size = 12),
                legend.text = element_text(family = 'Helvetica', size = 12),
                legend.title = element_text(family = 'Helvetica', size = 12),
                axis.line = element_blank(),
                strip.text = element_text(family = 'Helvetica', size = 12, face = 'bold'),
                strip.background = element_blank(),))

figure.dir.1 <-  'figures/'
```

## Gene sets for filtering, identification
```{r}
lnc.genes <- read.table('../data/gen.32.pure.lnc.names') %>% rename(gene = 'V1')
gencode.genes <- read.table('../data/gen.32.coding.names') %>% rename(gene = 'V1')

znfs.for.figure <- c('ZFP92', 'ZNF135', 'ZNF334', 'ZNF34', 'ZNF347', 'ZNF345A', 'ZNF345C',
                     'ZNF382', 'ZNF415', 'ZNF433', 'ZNF441', 'ZNF442', 'ZNF471', 'ZNF506',
                     'ZNF528', 'ZNF558', 'ZNF563', 'ZNF568', 'ZNF582', 'ZNF586', 'ZNF605',
                     'ZNF655', 'ZNF662', 'ZNF667', 'ZNF682', 'ZNF736', 'ZNF788', 'ZNF793',
                     'ZNF814', 'ZNF83', 'ZNF880', 'ZNF90', 'ZNF10', 'ZNF133', 'ZNF665', 'ZNF699',
                     'ZNF792')

ifn.signature <- c("DDX58", "IFIH1", "IFNL1", "IRF9", "IFI27", "OAS2", "IFIT3", "IFIT1", "OASL", "ISG15", "OAS3", "RSAD2", "MX1", "OAS1", "IFI44", "CH25H", "IFIT2", "ISG20", "RNASEL", "IFI6", "HERC5", "HERC6", "GBP3", "GBP1")

te.signature <- c('L1MEc', 'THE1D', 'L1MD2', 'MER20', 'L1MC4a', 'AluSz', 'AluY', 'AluSx1')

irds.gene.set <- c('ALDH3A1', 'BST2', 'CA2', 
                'CCNA1', 'CD59', 'CXCL1',
                'DAZ1', 'DCN', 'FLJ20035',
                'FLJ38348', 'G1P2', 'G1P3',
                'GALC', 'HERC6', 'HLA-B',
                'HLA-G', 'HSD17B1', 'IFI27',
                'IFI35', 'IFI44', 'IFI44L',
                'IFIT1', 'IFIT3', 'IFITM1',
                'IGF2', 'IRF7', 'LAMP3', 
                'LGALS3BP', 'LY6E', 'MCL1', 
                'MX1', 'MX2', 'OAS1', 'OAS3',
                'OASL', 'PLSCR1', 'RAP2C', 
                'ROBO1', 'SERPINB2', 'SH3YL1', 
                'SLC6A15', 'STAT1', 'THBS1', 
                'TIMP3', 'TncRNA', 'TRIM14', 
                'USP18', 'ZNF273')

```

## Normalized Counts (DESeqII Normalized)
```{r}
aale.normalized.counts <- 
  read_csv('../data/aale.exo.data/normalized.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  filter(! grepl(c(')n'), gene),
         ! grepl(c('^MT-'), gene))
plasma.normalized.counts <- 
  read_csv('../data/patient.data/normalized.counts.csv') %>% 
  rename(gene = 'X1') %>%  
  filter(! grepl(c(')n'), gene),
         ! grepl(c('^MT-'), gene))
h358.normalized.counts <- 
  read_csv('../data/reem.exo.rna.seq/normalized.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  filter(! grepl(c(')n'), gene),
         ! grepl(c('^MT-'), gene))
intra.aale.normalized.counts <- 
  read_csv('../data/in.vitro.data/normalized.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  filter(! grepl(c(')n'), gene),
         ! grepl(c('^MT-'), gene))

```

## Differential Expression
```{r}
aale.de.seq <- 
  read_csv('../data/aale.exo.data/de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1')
plasma.de.seq <- 
  read_csv('patient.data/de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') 
h358.de.seq <- 
  read_csv('../data/reem.exo.rna.seq/de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') 
new.intra.aale.de.seq <- 
  read_csv('../data/in.vitro.data/de.seq.csv') %>% 
  filter(pvalue < 0.05) %>% 
  rename(gene = 'X1') 
intra.aale.de.seq <- 
  read_csv('../data/aale.intra.kras.v.ctrl.de-seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') #%>% 
  filter(gene %in% lnc.genes$gene)
```

## Filtered DESeq data sets
```{r}
filtered.aale.de.seq <- 
  aale.de.seq %>% 
  top_n(30, wt = -padj)
filtered.h358.de.seq <- 
  h358.de.seq %>% 
  top_n(30, wt = log2FoldChange)
filtered.plasma.de.seq <- 
  plasma.de.seq %>%
  filter(grepl('^Alu', gene)) %>% 
  top_n(30, wt = -padj)
```

## Volcano Plots
```{r}
plot.volcano <- function(de.seq.in) {
  de.seq.in <- de.seq.in %>% 
    mutate(biotype = ifelse(gene %in% gencode.genes$gene, 'coding',
                            ifelse(gene %in% lnc.genes$gene, 'lncRNA', 'TE')))
  ggplot(de.seq.in, aes(y=-log10(padj),
                     x=log2FoldChange,
                     color = biotype)) +
    theme(axis.line = element_blank()) + 
  geom_point(size = 2, alpha = 0.5) + 
    geom_rug(size = 1, alpha = 0.3) + 
  scale_x_continuous(limits = c(-10.5, 10.5)) +
  #scale_y_continuous(limits = c(0, 12)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(pvalue)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) + 
  scale_color_manual(name='', values = viridis(3)) +
  ggrepel::geom_text_repel(aes(label = 
  ifelse(abs(-log10(padj)) > 1.3 & abs(log2FoldChange) > 5, 
                                        as.character(gene), '')),
                          box.padding = unit('0.2', 'lines'),
                          segment.color = 'gray',
                          color = 'black',
                          ylim = c(1.3, Inf), size = 4) 
}

plot.volcano(h358.de.seq)
```

## Shared Differential Expression
```{r}
plot.volcano <- function(de.seq.in, de.seq.filt) {
  de.seq.in <- de.seq.in %>% 
    mutate(biotype = ifelse(gene %in% gencode.genes$gene, 'coding',
                            ifelse(gene %in% lnc.genes$gene, 'lncRNA', 'TE'))) %>% 
    filter(gene %in% de.seq.filt$gene)
  ggplot(de.seq.in, aes(y=-log10(padj),
                     x=log2FoldChange,
                     color = biotype)) +
    theme(axis.line = element_blank()) + 
  geom_point(size = 2, alpha = 0.5) + 
    geom_rug(size = 1, alpha = 0.3) + 
  scale_x_continuous(limits = c(-10.5, 10.5)) +
  #scale_y_continuous(limits = c(0, 12)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(pvalue)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) + 
  scale_color_manual(name='', values = viridis(3)) +
  ggrepel::geom_text_repel(aes(label = 
  ifelse(abs(-log10(padj)) > 1.3 & abs(log2FoldChange) > 5, 
                                        as.character(gene), '')),
                          box.padding = unit('0.2', 'lines'),
                          segment.color = 'gray',
                          color = 'black',
                          ylim = c(1.3, Inf), size = 4) 
}

plot.de.scatter <- function(de.seq.in, de.seq.filt) {
  de.seq.in <- de.seq.in %>% 
    mutate(biotype = ifelse(gene %in% gencode.genes$gene, 'coding',
                            ifelse(gene %in% lnc.genes$gene, 'lncRNA', 
                                   'TE'))) 
  de.seq.merge <- merge(de.seq.in, de.seq.filt, by = 'gene')
  ggplot(de.seq.merge, aes(log2FoldChange.x, log2FoldChange.y, label = gene)) + 
    geom_text() + 
  geom_hline(yintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4)
}

plot.volcano(plasma.de.seq, intra.aale.de.seq)
plot.de.scatter(plasma.de.seq, h358.de.seq)
```

## Quick Heatmap of filtered Normalized Counts
```{r}
ifn.11 <- c(
'DDX58',
'IFNL1',
'IRF9',
'OAS2',
'CH25H',
'RNASEL',
'IFI6',
'HERC5',
'GBP3',
'GBP1',
'IRF1')

aale.combined.counts <- 
  read_csv('../data/aale.combined.quants/intra.exo.aale.normalized.counts') %>% 
  rename(gene = 'X1')
aale.combined.counts %>% 
#  merge(h358.normalized.counts, by = 'gene') %>% 
#h358.normalized.counts %>% 
  #filter(gene %in% gencode.genes$gene, gene %in% filter(h358.de.seq, abs(log2FoldChange) > 3)$gene) %>% 
  #filter_at(vars(-gene), all_vars(. > 1)) %>% 
  filter(gene %in% ifn.signature) %>% 
  #mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>% 
  #filter(grepl('', gene)) %>% 
  #top_n(20, wt = -gene.cv) %>% 
  #select(-gene.cv) %>% 
  gather(patient, count, -gene) %>% 
  mutate(count = log2(count + 1)) %>% 
  #top_n(20, wt = count) %>% 
  spread(patient, count) %>% 
  column_to_rownames('gene') %>% 
  pheatmap(color = viridis(400))

merge(intra.aale.normalized.counts, aale.normalized.counts, by = 'gene') %>% 
  filter(gene %in% ifn.11) %>% 
  gather(patient, count, -gene) %>%
  mutate(count = log2(count + 1)) %>% 
  mutate(count = scale(count)) %>% 
  spread(patient, count) %>% 
  column_to_rownames('gene') %>% 
  pheatmap(color = viridis(16))
```

## 'Gene Set' Correlation
```{r}
intra.aale.normalized.counts %>% 
  filter(gene %in% ifn.signature) %>% 
  gather(patient, count, -gene) %>% 
  mutate(set = ifelse(gene %in% ifn.signature, 'ifn', 'adar')) %>% 
  group_by(patient, set) %>% 
  summarize(avg = mean(count)) %>% 
  spread(set, avg) %>% 
  separate(patient, sep = '[.]', into = c('pat', 'condition', 'rep')) %>% 
  mutate(score = adar) %>% 
  ggplot(aes(condition, score, label = rep)) + 
  geom_boxplot(color = 'grey') + 
  geom_text(size=6)
```

## Venn Diagram -- Detection:
```{r}
plasma.kras.detected.genes <- 
  plasma.normalized.counts %>% 
  select(contains('kras'), gene) %>% 
  filter_at(vars(-gene), all_vars(. > 10)) %>% 
  select(gene)
aale.kras.detected.genes <- 
  aale.normalized.counts %>% 
  select(contains('kras'), gene) %>% 
  filter_at(vars(-gene), all_vars(. > 10)) %>% 
  select(gene)
intra.aale.upregulated.genes <- 
  new.intra.aale.de.seq %>% 
  filter(log2FoldChange > 1,
            padj < 0.01) %>%
  select(gene)


intersect(intra.aale.upregulated.genes$gene, plasma.kras.detected.genes$gene) %>% intersect(., aale.kras.detected.genes$gene)

temp.vd <- venn.diagram(
  x = list(
    plasma.kras.detected.genes %>% unlist() , 
    aale.kras.detected.genes %>% unlist() , 
    intra.aale.upregulated.genes %>% unlist()),
  category.names = c("plasma" , "ex-AALE" , "upreg-intra-AALE"),
  filename = NULL,
  output = TRUE ,
          imagetype = 'png',
          height = 480 , 
          width = 520 , 
          resolution = 300,
            compression = "lzw",
          lwd = 1,
          col=viridis(3),
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
          cex = 3,
          fontfamily = "Helvetica",
          cat.cex = 2,
          cat.default.pos = "outer",
          cat.pos = c(-27, 27, 135),
          cat.dist = c(0.055, 0.055, 0.085),
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
          rotation = 1
        )

grid.draw(temp.vd)

library(grDevices)

pdf(file=paste(figure.dir.1, "new.align.above.10.venn.pdf", sep = '/'))
    grid.draw(temp.vd)
dev.off()
```

## RNA Editing:
```{r}
ipsc.editing <- read_csv('../data/ipsc.alu.editing.out.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'intracellular')

ex.ipsc.editing <- read_csv('../data/alu.ipsc.edit.out.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'extracellular')

eb.ipsc.editing <- read_csv('../data/eb.alu.editing.out.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'eb intracellular')

rbind(ipsc.editing, ex.ipsc.editing) %>% 
  #rbind(eb.ipsc.editing) %>% 
  filter(condition == 'ctrl') %>% 
  ggplot(aes(experiment, signal)) + 
  geom_boxplot(position = position_dodge(width = 1)) + 
  geom_point(position = position_dodge(width = 1)) + 
  stat_compare_means(method = 't.test') +
  scale_color_manual(values = c(viridis(1), viridis(3)[2])) + 
  scale_y_continuous(limits = c(0,1.5)) + 
  xlab('') + 
  ylab('A2GEditingIndex') + 
  theme(axis.text.x = element_text(family = 'Helvetica', size = 10, face = 'bold'))

ggsave('figures/ipsc.ctrl.editing.2.5x2.5.pdf', 
       height = 2.5, 
       width = 2.5, 
       units = 'in')


aale.ex.editing <- read_csv('../data/aale.exrna.sx.sz.edit.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('patient','condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'aale exRNA')

aale.in.editing <- read_csv('../data/sz.sx.etc.aale.intra.edit.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'szsx intracellular')

patient.editing <- read_csv('../data/patient.exrna.sx.sz.edit.csv') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('patient','condition', 'rep')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'lung cancer plasma')

exotic.rna.editing <- read_csv('../data/reem.sx.sz.rna.edit.out') %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  separate(Sample, sep = '[.]', into = c('h358','condition', 'rep', 'out')) %>% 
  select(condition, rep, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  mutate(experiment = 'exoTICvRNeasy')


#rbind(aale.ex.editing, aale.in.editing) %>% 
rbind(exotic.rna.editing) %>% 
  ggplot(aes(condition, signal)) + 
  geom_boxplot(position = position_dodge(width = 1)) + 
  geom_point(position = position_dodge(width = 1)) + 
  stat_compare_means(method = 't.test', label.y = 1.4) + 
  scale_color_manual(values = c(viridis(1), viridis(3)[2])) + 
  scale_y_continuous(limits = c(0,1.4))


```

## PCA
```{r}
### Gencode PCA
tx.condition.mapping.volume.pca <-  
  aale.normalized.counts %>%
  filter(!grepl('^MT-', gene)) %>% 
  #filter(gene %in% c(filtered.plasma.de.seq$gene, ifn.signature)) %>% 
  column_to_rownames('gene') %>% 
  t() %>% 
  as.data.frame() %>% 
  prcomp(scale=F)

#tx.condition.mapping.volume.pca <-  
#  patient.meds %>%
#  column_to_rownames('patient.id') %>% 
#  prcomp(scale=F)

summary(tx.condition.mapping.volume.pca)
pcs.of.interest <- apply(tx.condition.mapping.volume.pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[1]
 
# convert output to dataframe
pca.out <- as.data.frame(tx.condition.mapping.volume.pca$x)
pca.out <- 
  pca.out %>% 
  rownames_to_column('patient.id') %>% 
  separate(patient.id, sep = '[.]', into = c('pat', 'condition', 'rep'))
# summarize PC contributions
pca.out.summary <- 
  as.data.frame(summary(tx.condition.mapping.volume.pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)
# summarize PC contributions
pca.out.contributions <- 
  as.data.frame(tx.condition.mapping.volume.pca$rotation) %>% 
  rownames_to_column('tx') %>% 
  gather(pc, contrib, -tx) %>% 
  #merge(gencode.reference,
  #      by.x = 'tx',
  #      by.y = 'gene.name') %>% 
  group_by(pc) %>% 
  mutate(contrib = abs(contrib)) %>% 
  filter(contrib > quantile(contrib, 0.95))

# plot PC1 and PC2
ggplot(pca.out,
       aes(PC1, PC2, 
           color = condition,
           label = paste(paste(condition,rep, sep = '.'),
                         sep = ', '))) + 
  geom_point(alpha = 4) + 
  scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                          digits = 3), sep = ' ')) +
 scale_color_viridis_d(direction = -1, name = 'condition') + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  ggrepel::geom_text_repel(size = 4,
                            point.padding = 0.2,
                            box.padding = 0.11,
                            color = 'black',
                           segment.colour = NA) + 
  theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(0.4, 'cm'))

#ggsave("lnc.pca.plot.2.5x1.5.legend.pdf", width = 2.5, height = 1.5, units = 'in')

# print top contibutors per PC
pca.out.contributions %>%
  group_by(pc) %>%
  filter(pc %in% c("PC1", "PC2")) %>% 
  select(pc, contrib, tx) %>% 
  mutate(contrib = abs(contrib)) %>% 
  top_n(10, wt = contrib) %>% 
  arrange(pc, contrib)
```

## Summary Figures:

### Transcript Summary Viz

#### Normalized read count dist
```{r}
tx.count.log2.counts.dist.plot <- 
  salmon.tx.normalized.counts.tidy %>% 
  filter(normalized.count > 0) %>% 
  #merge(patient.info.encoded[,c(4,6)], 
  #      by = 'patient.id') %>% 
  filter(biotype %in% c('protein-coding',
                        'lncRNA')) %>% 
  group_by(tx, patient.id) %>% 
  mutate(condition.patient = paste(condition, patient, sep = '.')) #%>%
  #filter(condition.patient == 'ctrl.1') %>%
   #%>% 

ggplot(tx.count.log2.counts.dist.plot,
       aes(reorder(condition.patient, age), 
             log2(as.numeric(normalized.count) + 1),
             color = condition)) + 
  geom_boxplot(alpha = 0.4, position = 'dodge') + 
  scale_y_continuous(limits = c(0,15)) + 
  #scale_color_manual(name = '', values = c('#37b578ff',
  #                                         '#440154ff')) + 
  scale_color_npg() +
  facet_wrap(~biotype, ncol = 2) +
  xlab('') 

tx.count.log2.counts.dist.plot

```

#### Unique lnc Tx bar
```{r}
tx.unique.lncrna.count <- 
  salmon.tx.normalized.counts.tidy %>% 
  mutate(condition.patient = paste(condition, patient, sep = '.')) %>% 
  filter(normalized.count > 1, 
         biotype %in% c('protein-coding', 'lncRNA')) %>% 
  group_by(patient.id, input_volume, condition, biotype, age) %>% 
  #filter(condition.patient == 'ctrl.1') %>%
  summarise(count = n_distinct(tx)) %>% 
  ggplot(aes(reorder(patient.id, age), 
             count, 
             fill = age)) + 
  geom_col(position = 'dodge') + 
  #coord_flip() +
  geom_hline(yintercept = seq(0, 15000, 2500), color = 'white') + 
  scale_fill_viridis_c(direction = -1) + 
  xlab('') + 
  ylab('unique transcripts') + 
  facet_wrap(~biotype, nrow = 1) + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1))

tx.unique.lncrna.count
```

#### Stacked unique Tx per chromosome
```{r}
chr.size <-  
  salmon.tx.normalized.counts.tidy %>% 
  group_by(chr, condition) %>% 
  summarise(total.chr.counts = sum(as.numeric(normalized.count)))

stacked.unique.tx.by.chrom <- 
  salmon.tx.normalized.counts.tidy %>% 
  mutate(condition.patient = paste(condition, patient, sep = '.')) %>% 
  filter(normalized.count > 1) %>% 
  group_by(condition.patient, chr ,condition, biotype) %>% 
  #filter(grepl('^panc', condition.patient)) %>
  filter(biotype == 'lncRNA') %>% 
  summarise(count = n_distinct(tx)) %>% 
  merge(chr.size, 
        by = c('chr', 'condition')) %>% 
  group_by(condition, chr, biotype) %>% 
  summarize(mean.count = mean(count)) %>% 
  ggplot(aes(reorder(chr, mean.count),
             mean.count,
             fill = ),
         alpha = 0.6) + 
  geom_col(position = 'stack') + 
  #coord_flip() +
  scale_fill_viridis_d(direction = -1) + 
  xlab('Chromosome') + 
  ylab('unique transcripts') + 
  facet_wrap(~condition, nrow = 1) + 
  scale_fill_manual(name = '', values = c('#37b578ff',
                                           '#440154ff'))
stacked.unique.tx.by.chrom
```

#### Summary Table
```{r}
salmon.tx.normalized.counts.tidy %>% 
  group_by(mapping.rate, condition, patient, input_volume) %>% 
  filter(NumReads > 5) %>% 
  summarise(tot.num.reads = round(sum(NumReads)),
            tx.count = n_distinct(tx)) %>% 
  mutate(approx.tot.reads = round(tot.num.reads/mapping.rate)) %>% 
  ungroup() %>% 
  group_by(condition) %>% 
  arrange(patient) %>% 
  select(`Patient` = 'patient',
         `Input Volume (mL)` = 'input_volume',
         `Mapping Rate` = 'mapping.rate',
         `Condition` = 'condition',
         `Total Aligned Reads` = 'tot.num.reads',
         `Distinct Transcripts` = 'tx.count',
         `Approximate Total Reads` = 'approx.tot.reads') %>% 
  gt() %>% 
  tab_header(
    title = 'Table 1.',
    subtitle = 'Summary of library psuedoalignments'
  ) %>% 
  fmt_number(
    columns = vars(`Input Volume (mL)`,
                   `Total Aligned Reads`, 
                   `Distinct Transcripts`, 
                   `Approximate Total Reads`),
    drop_trailing_zeros = T) %>% 
  fmt_percent(
    columns = vars(`Mapping Rate`)) %>% 
  tab_stubhead(label = 'Sample')

```

## Gencode PCA and Correlation
```{r}
# prepare for matrix-style input
# filter out genes with low coefficient of variance and 
# genes that are present on chrY,M and Hemoglobin 

salmon.tx.norm <- 
  salmon.normalized.counts %>% 
  #filter(! gene %in% chrM.txs$gene.name) %>% 
  column_to_rownames('gene')

salmon.tx.norm.collison <- 
  salmon.normalized.counts %>% 
  filter(gene %in% collison.subtype.genes$Genes) %>% 
  column_to_rownames('gene')

ctrl.consistent.genes <- 
  as.data.frame(t(salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  filter(grepl('ctrl', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  #filter(gene %in% lnc.genes$gene.name) %>% 
  filter_at(vars(-gene), all_vars(. > 1)) %>% 
  #subset(rowMeans(.[,-c(1)]) > 13) %>% 
  mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>% 
  #filter(gene.cv > quantile(gene.cv, 0.70, na.rm = T)) %>%
  filter(gene.cv < quantile(gene.cv, 0.2, na.rm = T)) %>%
  select(-gene.cv)

panc.consistent.genes <- 
  salmon.tx.norm.t <- 
  as.data.frame(t(salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  filter(grepl('panc', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  filter(gene %in% lnc.genes$gene.name) %>% 
  filter_at(vars(-gene), all_vars(. > 1)) %>% 
  #subset(rowMeans(.[,-c(1)]) > 13) %>% 
  mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>% 
  #filter(gene.cv > quantile(gene.cv, 0.50, na.rm = T),
  #       ! gene %in% ctrl.consistent.genes$gene) %>%
  filter(gene.cv < quantile(gene.cv, 0.2, na.rm = T),
         ! gene %in% ctrl.consistent.genes$gene) %>%
  select(-gene.cv)

# filter to DE genes
#de.genes.list <-  
#  grouped.de.seq %>% 
#  filter(abs(log2FoldChange) > 0)


# transpose data frame and add metadata while applying filters of choice
salmon.tx.norm.t <- 
  as.data.frame(t(salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  #filter(grepl('panc', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  filter_at(vars(-gene), any_vars(. > 0)) %>% 
  subset(rowMeans(.[,-c(1)]) > 13) %>% 
  filter(gene %in% ctrl.consistent.genes$gene) %>% 
  #filter(gene %in% c(panc.consistent.genes$gene,
  #                   ctrl.consistent.genes$gene)) %>% 
  remove_rownames() %>% 
  column_to_rownames('gene') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('patient.id') %>% 
  merge(patient.info.encoded,
        by = 'patient.id') %>% 
  #left_join(patient.meds)  %>% 
  column_to_rownames('patient.id') 
```

### Gencode Correlation Matrices
```{r}
# plot correlation matrix sample vs. sample
ggcorrplot(cor(t(salmon.tx.norm.t)),
           hc.order = T, lab = F) + 
  scale_fill_viridis_c(direction = -1,
                       limits = c(-0.15, 1), option = 'viridis',
                       name = 'Pearson') + 
  ylab('') + 
  xlab('') +
            theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(0.4, 'cm'),
          panel.background = element_rect(fill = 'white'),
                axis.text.x = element_text(family = 'Helvetica', size = 6),
                axis.text.y = element_text(family = 'Helvetica', size = 6),
                axis.title.x = element_text(family = 'Helvetica', size = 6),
                axis.title.x.bottom = element_text(family = 'Helvetica', size = 6),
                axis.title.y = element_text(family = 'Helvetica', size = 6),
                legend.text = element_text(family = 'Helvetica', size = 6),
                legend.title = element_text(family = 'Helvetica', size = 6),
                axis.line = element_blank(),
                strip.text = element_text(family = 'Helvetica', 
                                          size = 6),
                strip.background = element_blank())

#ggsave("corr.pdf", width = 2.5, height = 1.5, units = 'in')
```



### Elbow Plot Fxn:
```{r}
elbow.plt <- function(df){
  k.max <- nrow(df) - 1
  k.values <- 1:k.max
  
  cluster.compute <- function(k){
    cluster.out <- kmeans(df, 
                          k, 
                          nstart = 25)
    return(cluster.out$tot.withinss)
  }
  
  elbow.df <- tibble()
  for(i in k.values){
    print(i)
    temp.df <- 
      tibble(k = i, 
             ss = cluster.compute(i))
    elbow.df <- rbind(temp.df, elbow.df)
  }
  
  ggplot(elbow.df,
         aes(k, ss)) + 
    geom_point() +
    geom_line() + 
    geom_vline(xintercept = k.values, 
               linetype = 'dashed', 
               alpha = 0.3)
}
```

### Gencode Kmeans
```{r}
set.seed(234)

hc.pca <- hclust(as.dist(pca.out), method = "ward.D2")

pca.for.clustering <- 
  pca.out %>% 
  select(PC1, PC2)

elbow.plt(pca.for.clustering)

kmeans(pca.for.clustering, 3) -> 
  pca.k.means

pca.for.clustering %>% 
  rownames_to_column('patient.name') %>% 
  mutate(cluster = factor(pca.k.means$cluster)) %>%  
  merge(patient.info.encoded,
        by.x = 'patient.name',
        by.y = 'patient.id') %>%  
  merge(salmon.tx.normalized.counts.tidy[,c(1,19)], 
        by.x = 'patient.name',
        by.y = 'patient.id') %>% 
  distinct() %>% 
  mutate(condition = ifelse(condition == 0 , 'ctrl', 'panc')) ->
  pca.for.clustering

pca.for.clustering %>% 
  ggplot(aes(PC1, 
             PC2, 
             color = cluster, 
             label = paste(paste(condition,patient, sep = '.'),
                           #paste0(input_volume,'mL'),
                           paste0(age,'yrs'),
                           sep = ', '))) + 
  geom_point() + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                          digits = 3), sep = ' ')) +
  scale_color_viridis_d() + 
  ggrepel::geom_text_repel(size = 1.5,
                           point.padding = 0.1,
                          box.padding = 0.2,
                          segment.color = NA,
                          color = 'black') +
  
  theme(legend.position = 'right')
ggsave("clustered.pca.plot.5x1.5.legend.pdf", width = 5, height = 1.5, units = 'in')


panc.salmon.tx.norm.t <- 
  salmon.tx.norm.t %>% 
  rownames_to_column('patient') %>% 
  filter(!grepl('ctrl', patient)) %>% 
  column_to_rownames('patient')
  
hc <- hclust(as.dist(salmon.tx.norm.t), method = "ward.D2")

plot(as.dendrogram(hc), 
     type = 'rectangle', 
     ylab = 'Height')

hc.dendro <- dendro_data(hc, type = 'rectangle')
ggplot() +
  geom_segment(data=segment(hc.dendro), 
               aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_text(data=label(hc.dendro), 
            aes(x=x, y=y, label=label, hjust=-0.15), size=3) +
  coord_flip() + 
  ylab('Height') + 
  scale_y_reverse(expand=c(0.2, 0)) + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

plot(as.dendrogram(hc.pca), 
     type = 'rectangle', 
     ylab = 'Height')
```

### Gencode UMAP
```{r}
set.seed(234)
uwot.gencode <- as.data.frame(umap(salmon.tx.norm.t,
     n_neighbors = 3.605551, learning_rate = 0.5))

rownames(uwot.gencode) <- rownames(salmon.tx.norm.t)

uwot.gencode %>% 
  rownames_to_column('patient.id') %>% 
  merge(patient.info.encoded, by = 'patient.id') %>% 
  mutate(condition = ifelse(condition == 1, 'Panc', 'Ctrl')) %>% 
  merge(salmon.tx.normalized.counts.tidy[,c(1,19)],
        by = 'patient.id') %>% 
  distinct() -> uwot.gencode.plot

ggplot(uwot.gencode.plot, 
       aes(V1, 
           V2, 
           color = condition,
           label = patient)) +
  #geom_text(size = 1) + 
  ylab('U2') + 
  xlab('U1') +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  ggrepel::geom_text_repel(size = 1,
                           segment.color = NA,
                           point.padding = NA)  + 
 scale_color_manual(name = 'condition', 
                    values = c('#37b578ff', '#440154ff')) + 
  theme(legend.position = 'right',
        legend.spacing = unit(0, 'in'))


elbow.plt(uwot.gencode)

kmeans(uwot.gencode, 3) -> 
  uwot.k.means

uwot.gencode %>% 
  rownames_to_column('patient') %>% 
  mutate(cluster = factor(uwot.k.means$cluster)) %>%  
  merge(patient.info.encoded,
        by.x = 'patient',
        by.y = 'patient.id') %>%  
  mutate(condition = ifelse(condition == 0 , 'ctrl', 'panc')) ->
  uwot.for.clustering

uwot.for.clustering %>% 
  ggplot(aes(V1, 
             V2, 
             color = cluster, 
             label = paste(patient,
                           paste0(input_volume,'mL'),
                           paste0(age,'yrs'),
                           sep = '\n'))) + 
  geom_point() + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.6) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.6) +
  ggrepel::geom_label_repel(box.padding = 0.5,
                           segment.alpha = 0.3) +
  scale_color_()


```

#### Gencode UMAP w/ Genes
```{r}
uwot.genes <- as.data.frame(umap(as.data.frame(t(salmon.tx.norm.t)),
     n_neighbors = sqrt(ncol(salmon.tx.norm.t)), learning_rate = 0.5, pca = 50))
ggplot(uwot.genes, 
       aes(V1, 
           V2)) + 
  geom_hline(yintercept = 0, linetype = 'dotted') + 
  geom_vline(xintercept = 0, linetype = 'dotted') +
  geom_point()

```

### Gencode NMF
```{r}
salmon.tx.norm.for.nmf <- 
  salmon.tx.norm.t %>% 
  select(-condition)
  
salmon.tx.norm.for.nmf <- 
  salmon.tx.norm.for.nmf[, colSums(salmon.tx.norm.for.nmf) != 0] %>% 
  t() %>% 
  as.data.frame()

for(n in seq(2,5)){
  set.seed(123)
  nmf.out <- nmf(salmon.tx.norm.for.nmf, n, nrun = 10)
  fit(nmf.out)
  print(paste(n, cophcor(nmf.out), sep = ':'))
}

nmf.out <- nmf(salmon.tx.norm.for.nmf, 3, nrun = 10)

as.data.frame(as.table(featureScore(nmf.out)[extractFeatures(nmf.out, 30L)[[1]]])) #%>% 
  merge(collison.subtype.genes,
        by.x = 'Var1',
        by.y = 'Genes') %>% 
  mutate(group = 'one') %>% 
  gather(type, score, -Var1, -group, -Freq) %>% 
  mutate(weighted.score = Freq * score) %>% 
  group_by(group, type) %>% 
  summarise(tot.score = mean(weighted.score) * 100)  -> group.1.nmf

as.data.frame(as.table(featureScore(nmf.out)[extractFeatures(nmf.out, 10L)[[2]]])) #%>% 
  merge(collison.subtype.genes,
        by.x = 'Var1',
        by.y = 'Genes') %>% 
  mutate(group = 'two') %>% 
  gather(type, score, -Var1, -group, -Freq) %>% 
  mutate(weighted.score = Freq * score) %>% 
  group_by(group, type) %>% 
  summarise(tot.score = mean(weighted.score) * 100) -> group.2.nmf


qqnorm(featureScore(nmf.out))
qqline(featureScore(nmf.out))
basismap(nmf.out)
coefmap(nmf.out)
consensusmap(nmf.out)
```

### Plotting Subtypes
```{r}
collison.groups.nmf <- rbind(group.1.nmf, 
                             group.2.nmf)
                             #group.3.nmf,
                             #group.4.nmf)

ggplot(collison.groups.nmf,
       aes(group, tot.score, fill = type)) + 
  geom_col(position = 'dodge') + 
  scale_fill_viridis_d() + 
  xlab('') + 
  ylab('Specificity Score') + 
  geom_hline(yintercept = 
               seq(0,max(collison.groups.nmf$tot.score),20), color = 'white')
```

### Group-based DEseq
```{r}
group.1.de.seq <- 
  read_csv('data/7.6.5.10.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 3')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')
  
group.2.de.seq <- 
  read_csv('data/8.3.4.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 2')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')

group.3.de.seq <- 
  read_csv('data/9.2.1.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 4')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')
overall.de.seq <- 
  read_csv('data/overall.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct()


grouped.de.seq <- bind_rows(group.1.de.seq, 
                            group.2.de.seq,
                            group.3.de.seq)

grouped.de.seq.for.table <- 
  grouped.de.seq %>% 
  merge(gencode.reference,
        by.x = 'gene',
        by.y = 'gene.name') %>% 
  mutate(genome = 'Hg38') %>% 
  select(enst='tx', chromosome='chr', start.position='start',stop.position='stop',
         strand, transcript.id='tx.name', gene, len='tx.len', log2FoldChange,
         padj, biotype, genome) %>% 
  distinct()
grouped.de.seq.for.table.lncs <- 
  grouped.de.seq.for.table %>% 
  filter(biotype == 'lncRNA')
grouped.de.seq.for.table.coding <- 
  grouped.de.seq.for.table %>% 
  filter(biotype != 'lncRNA')

grouped.de.seq %>% 
  group_by(group) %>% 
  mutate(biotype = ifelse(gene %in% lnc.genes$gene.name, 'lncRNA', 'Coding')) %>% 
  top_n(10, wt = log2FoldChange) %>% 
  ggplot(aes(log2FoldChange, 
             -log10(padj),
             label = gene,
             color = biotype,
             shape = group)) + 
    geom_point() +
  scale_color_viridis_d() + 
  ggrepel::geom_text_repel(size = 1, 
                           color = 'black', 
                           segment.colour = NA) +
      theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(0.4, 'cm')) 


 
grouped.de.seq %>% 
  merge(collison.subtype.genes,
        by.x = 'gene',
        by.y = 'Genes') %>% 
  ggplot(aes(gene, log2FoldChange)) + 
  geom_col() + 
  facet_wrap(~group, scales = 'free_x') + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1)) + 
  ylab('')

  

#ggsave("summary.de.pdf", width = 2.5, height = 1.5, units = 'in')


```

#### Viz
```{r}
de.seq.plot <- function(de.data){
  de.data %>% 
    mutate(biotype = ifelse(gene %in% lnc.genes$gene.name, 'lncRNA', 'Coding')) %>%
ggplot(aes(log2FoldChange, 
             -log10(padj), 
             color = biotype)) +
  geom_point(alpha = 0.75, 
             size = 1) + 
  geom_rug(alpha = 0.2) +
  geom_vline(xintercept = 0, 
             color = 'black', 
             linetype = 'dotted') +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') + 
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', 
             alpha = 0.4, 
             size = 0.2) +
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) +
  geom_vline(xintercept = c(-1,1), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) + 
    scale_color_manual(name = '', values = c('#440154ff',
                                           '#37b578ff')) + 
    theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(1, 'cm')) + 
    geom_text_repel(aes(label = ifelse(log2FoldChange > 3 & -log10(padj) > 3 & biotype == 'lncRNA', 
                                   gene, '')),
                    segment.color = NA,
                    point.padding = NA) 
}

de.seq.plot(overall.de.seq)
ggsave('overall.de.seq.lncs.label.pdf', height = 10, width = 25, units = 'in')

de.seq.plot <- function(de.data){
  de.data %>%
    filter(abs(log2FoldChange) > 1) %>% 
    mutate(biotype = ifelse(gene %in% lnc.genes$gene.name, 'lncRNA', 'Coding')) %>% 
    rename(FDR = padj) %>% 
    group_by(biotype) %>% 
  ggplot(aes(log2FoldChange,
             -log10(FDR),
             label = gene,
             color = biotype)) + 
    geom_point(size = 0.5, alpha = 0.55) +
    geom_vline(xintercept = 0, linetype = 'dotted', size = 0.5) +
    xlab('') + scale_color_manual(name = '', values = c('#440154ff',
                                           '#37b578ff')) + 
    #scale_color_viridis_d() + #guide = guide_colorbar(reverse = T)) + 
    #geom_text_repel(aes(label = ifelse(log2FoldChange > 11, 
    #                               gene, '')),
    #                segment.color = NA,
    #                point.padding = NA) +
    #scale_y_continuous(limits = c(-13,13)) + 
    facet_wrap(~group+biotype, ncol = 2) + 
    theme(legend.position = 'bottom',
          legend.key.width = unit(0.6,'cm'),
          legend.key.height = unit(0.125, 'cm')) #+ 
    #coord_flip()
}

ggsave('de.log2.plot.3x1.5.pdf', height = 3, width = 2.5, units = 'in')


de.seq.plot(group.1.de.seq)
de.seq.plot(group.2.de.seq)
de.seq.plot(group.3.de.seq)


merge(group.2.de.seq, 
      group.3.de.seq,
      by = 'gene') %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, label = gene)) + 
  geom_text()

```

## FGSEA
```{r}
de.seq.rank <- 
  group.3.de.seq %>% 
  mutate(rank = sin(log2FoldChange) * -log10(padj + 0.0000001),
         rank = as.double(rank),
         gene = as.character(gene)) %>% 
  select(gene, rank) %>%
  drop_na() %>% 
  arrange(-rank) 

clipr::write_clip(filter(group.1.de.seq, log2FoldChange > 3)$gene)

aseq.rank <- de.seq.rank$rank
seq.rank <- setNames(de.seq.rank$rank, de.seq.rank$gene)

all.msig.df <- 
  msigdbr(species = 'Homo sapiens') %>% 
  select(gene_symbol, gs_name) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

hallmark.msig.df <- 
  msigdbr <- 
    msigdbr(species = 'Homo sapiens', category = 'H') %>% 
   select(gene_symbol, gs_name)%>% 
  split(x = .$gene_symbol, f = .$gs_name)

fgseaRes <- 
  fgsea(pathways = all.msig.df, 
        stats = seq.rank,
        nperm = 100000)


fgseaRes.hallmark <- 
  fgsea(pathways = hallmark.msig.df, 
        stats = seq.rank,
        nperm = 100000)


gsea.df <- as_tibble(fgseaRes) 
hm.gsea.df <- as_tibble(fgseaRes.hallmark)
```



## TCGA/GTEX Comparisons:
### TCGA
```{r}
tcga.paad.phenotypes <- 
  read_tsv('data/TCGA-PAAD.GDC_phenotype.tsv.gz')[,c(1,5,30,40)] %>% 
  rename( sample.id = 'submitter_id.samples')

tcga.paad.norm.counts <- 
  read_tsv('data/TCGA-PAAD.htseq_counts.tsv.gz') %>% 
  gather(sample.id, counts, -Ensembl_ID) %>%
  subset(sample.id %in% tcga.paad.phenotypes$sample.id) %>% 
  merge(tcga.paad.phenotypes, by.all = sample.id) %>% 
  add_column(source = 'TCGA-panc')

tcga.paad.norm.counts.phenotypes.matrix.fmt <- 
  tcga.paad.norm.counts %>% 
  select(patient = 'sample.id',
         Ensembl_ID,
         counts,
         source) %>% 
  merge(gencode.names,
        by.x = 'Ensembl_ID',
        by.y = 'ensg') %>% 
  select(patient, gene = 'gene.name', counts, source) %>% 
  mutate(match = str_split_fixed(patient, 
                                 pattern = '-', 
                                 n = 4)[,4],
         patient.tag = str_sub(str_split_fixed(patient, 
                                 pattern = '-', 
                                 n = 2)[,2], 1, 
                               7),
         source = ifelse(match == '11A', 
                         'TCGA-panc-normal',
                         'TCGA-panc-tumor')) %>% 
  select(-match) %>% 
  group_by(patient.tag, source)

match.normal.list <- 
  tcga.paad.norm.counts.phenotypes.matrix.fmt %>% 
  distinct(patient.tag, source) %>% 
  filter(source == 'TCGA-panc-normal')

matched.tcga.paad.mtx.fmt <- 
  tcga.paad.norm.counts.phenotypes.matrix.fmt %>% 
  ungroup() %>% 
  filter(patient.tag %in% match.normal.list$patient.tag) %>% 
  select(-patient.tag)


table(grepl('11A',tcga.paad.norm.counts.phenotypes.matrix.fmt$patient))
```
### GTEX:
```{r}
gtex.phenotypes <- 
  read_tsv('../KRAS.AALE/rna.seq.analysis/data/GTEX_phenotype.gz') %>% 
  select(`_primary_site`, Sample) %>% 
  filter(`_primary_site` %in% c("Pancreas"))

gtex.norm.counts <- 
  read_tsv('../KRAS.AALE/rna.seq.analysis/data/gtex.norm.counts.tsv.gz') %>% 
  gather(patient, counts, -sample)


gtex.norm.counts.phenotypes <- 
 gtex.norm.counts %>% 
    merge(gtex.phenotypes,
        by.x = 'patient',
        by.y = 'Sample')

gtex.col.data <- 
  gtex.norm.counts.phenotypes %>% 
  select(patient, condition = '_primary_site') %>% 
  distinct(patient, condition) %>%  
  column_to_rownames('patient')

gtex.norm.counts.phenotypes.matrix.fmt <- 
  gtex.norm.counts.phenotypes %>% 
  mutate(counts = 2^counts,
         counts = round(counts - 1)) %>% 
  select(-`_primary_site`) %>% 
  group_by(patient) %>% 
  spread(patient, counts) %>% 
  distinct(sample, .keep_all = T) %>% 
  column_to_rownames('sample')

all(rownames(gtex.col.data) == colnames(gtex.norm.counts.phenotypes.matrix.fmt))

gtex.dds <- DESeq2::DESeqDataSetFromMatrix(countData = gtex.norm.counts.phenotypes.matrix.fmt,
                                           colData = gtex.col.data,
                                           design = ~condition)

gtex.dds <- DESeq2::estimateSizeFactors(gtex.dds)

gtex.de.seq.norm.counts <- as.data.frame(DESeq2::counts(gtex.dds, normalized=T))


salmon.normalized.counts %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name)) %>% 
  gather(patient.id, norm.count, -gene) %>% 
  distinct(patient.id, gene, norm.count, .keep_all = T) %>% 
  spread(patient.id, norm.count) -> clean.salmon.normalized.counts

gtex.tissue.origin <- 
  gtex.col.data %>% 
  rownames_to_column('patient')

gtex.salmon.ready.counts <- 
  gtex.de.seq.norm.counts %>% 
  rownames_to_column('gene') %>% 
  gather(patient, counts, -gene) %>% 
  merge(gtex.tissue.origin,
        by = 'patient')
  #select(patient, sample = 'gene', counts)

gtex.salmon.merged.counts <- 
  clean.salmon.normalized.counts %>% 
  gather(patient, counts, -gene) %>% 
  mutate(condition = ifelse(grepl('^ctrl', patient), 
                                'ctrl.plasma', 'panc.plasma')) %>% 
  bind_rows(gtex.salmon.ready.counts) %>% 
  bind_rows(select(filter(tcga.paad.norm.counts.de.seq.norm.plasma,
                          origin != 'plasma'),
                   patient = 'patient.id', 
                   gene,
                   counts,
                   `_primary_site` = 'condition')) %>% 
  group_by(condition) %>% 
  mutate(counts = ifelse(is.na(counts), 0, counts)) %>% 
  filter(counts > quantile(counts, 0.95))

gtex.salmon.merged.counts.tissue.avg <-
  gtex.salmon.merged.counts %>% 
  #read.csv('test.cor.salmon.long.csv') %>% 
  rename(tissue = 'condition') %>% 
  group_by(tissue, gene) %>% 
  #select(-X) %>% 
  summarise(tissue.counts = mean(counts))

for.cor.gtex.salmon.merged.counts <- 
  gtex.salmon.merged.counts.tissue.avg %>% 
  group_by(tissue, gene) %>% 
  spread(gene, tissue.counts, fill = 0) %>% 
  column_to_rownames('tissue')

for.cor.gtex.salmon.merged.counts <- 
  for.cor.gtex.salmon.merged.counts[,colSums(for.cor.gtex.salmon.merged.counts) > 7]

write.csv(for.cor.gtex.salmon.merged.counts, 'cor.plasma.gtex.csv')

for.cor.gtex.salmon.merged.counts <- 
  read.csv('cor.plasma.gtex.csv') %>% 
  column_to_rownames('X')
  
#cor.gtex.salmon.merged.counts <- cor(for.cor.gtex.salmon.merged.counts)
cor.gtex.salmon.merged.counts <- cor(t(for.cor.gtex.salmon.merged.counts))
head(cor.gtex.salmon.merged.counts)

ggcorrplot(cor.gtex.salmon.merged.counts,
           hc.or <- der = T, lab = T, type = 'upper')


tcga.expression.pca <- prcomp(cor.gtex.salmon.merged.counts)
pca.out <- as.data.frame(tcga.expression.pca$x)
#pca.out$group <- as.character(gtex.tcga.plasma.cor$origin)
#pca.out$group <- ifelse(pca.out$group == 0 , 'TCGA', 
                        #ifelse(pca.out$group == 1,'plasma', 'GTEX'))


for.pca.gtex.salmon.merged.counts <- 
  gtex.salmon.merged.counts %>% 
  rename(tissue = '_primary_site') %>% 
  
multi.var.de.seq.w.ensg <- 
  multi.var.de.seq %>% 
  merge(gencode.reference[,c(6,8)],
        by.x = 'gene',
        by.y = 'gene.name') %>% 
  filter(padj < 0.01) %>% 
  distinct() %>% 
  select(1,3,6,8)

tcga.paad.norm.counts.de.seq <- 
  tcga.paad.norm.counts %>% 
  merge(multi.var.de.seq.w.ensg,
        by.x = 'Ensembl_ID',
        by.y = 'ensg') %>% 
  select(-4,-5,-6,-7,-8)

salmon.normalized.counts %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name)) %>% 
  gather(patient.id, norm.count, -gene) %>% 
  distinct(patient.id, gene, norm.count, .keep_all = T) %>% 
  spread(patient.id, norm.count) %>% 
  merge(gencode.reference[,c(6,8)],
        by.x = 'gene',
        by.y = 'gene.name') -> clean.salmon.normalized.counts
  
  
tcga.paad.norm.counts.de.seq.norm.plasma <- 
  tcga.paad.norm.counts.de.seq %>% 
  spread(sample.id, as.numeric(counts)) %>% 
  merge(clean.salmon.normalized.counts,
        by.x = 'Ensembl_ID',
        by.y = 'ensg') %>% 
  select(-Ensembl_ID) %>% 
  distinct() %>% 
  gather(patient.id, counts, -gene, -log2FoldChange, -padj) %>% 
  mutate(origin = ifelse(grepl('^TCGA', patient.id), 'TCGA', 'plasma')) %>% 
  #filter(origin == 'TCGA') %>% 
  #filter(log2FoldChange > 4) %>% 
  #filter(gene %in% lnc.rna.names$gene) %>% 
  group_by(gene) %>% 
  filter(mean(counts) > 3)

#tcga.paad.norm.counts.de.seq.norm.plasma %>% 
#  ggplot(aes(reorder(gene, log2FoldChange), counts, color = padj)) + 
#  geom_boxplot(alpha = 0.4) + 
#  scale_color_viridis_c() + coord_flip()

temp.tcga <- tcga.paad.norm.counts.de.seq.norm.plasma[, c(3:6)]

gtex.tcga.gene.merge <- 
  gtex.norm.counts %>% 
  filter(gene %in% tcga.paad.norm.counts.de.seq.norm.plasma$gene) %>% 
  bind_rows(temp.tcga) %>% 
  merge(tcga.paad.norm.counts.de.seq.norm.plasma[,c(1,2,3)], 
        by = 'gene') %>% 
  distinct() 

gtex.tcga.gene.merge.norm <- 
  gtex.tcga.gene.merge %>% 
  group_by(origin) %>% 
  mutate(norm.counts = scale(counts/sum(counts))) %>% 
  #filter(origin != 'GTEX') %>% 
  filter(log2FoldChange >= quantile(log2FoldChange, 0.85),
         padj <= 0.01) %>% 
  distinct()

ord <- 
  hclust(dist(gtex.tcga.gene.merge.norm$norm.counts, 
               method = "euclidean"), method = "ward.D" )$order


gtex.tcga.gene.merge.norm$patient.id <- 
  factor(gtex.tcga.gene.merge.norm$patient.id, 
         levels = gtex.tcga.gene.merge.norm$patient.id[ord])


gtex.tcga.gene.merge.norm %>% 
  select(gene, patient.id, norm.counts) %>% 
  #spread(patient.id, norm.counts) %>% 
  spread(gene, norm.counts, fill = 0) %>% 
  ungroup() %>% 
  #distinct(gene, .keep_all = T) %>% 
  mutate(origin = ifelse(origin == 'TCGA', 0, 
                         ifelse(origin == 'plasma', 1, 2))) %>% 
  column_to_rownames('patient.id') -> gtex.tcga.plasma.cor #%>% 
  cor() -> gtex.tcga.plasma.cor
  
tcga.expression.pca <- prcomp(gtex.tcga.plasma.cor)
pca.out <- as.data.frame(tcga.expression.pca$x)
pca.out$group <- as.character(gtex.tcga.plasma.cor$origin)
pca.out$group <- ifelse(pca.out$group == 0 , 'TCGA', 
                        ifelse(pca.out$group == 1,'plasma', 'GTEX'))

ggplot(pca.out,
       aes(PC1, PC2, 
           color = group,
           label = rownames(pca.out))) + 
  geom_point() + 
  scale_color_viridis_d(direction = -1)

ggcorrplot(gtex.tcga.plasma.cor)

ggplot(gtex.tcga.gene.merge.norm,
       aes(patient.id, gene[ord], fill = norm.counts)) + 
  geom_raster() + 
  scale_fill_distiller(palette = 'RdYlBu') + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))




```


## Coverage
```{r}
kras.panc.cov <- read_tsv('data/panc.kras.cov.tsv') %>% 
  rename(chr = `#'chr'`,
         start = `'start'`,
         end = `'end'`) %>% 
  gather(patient, counts, -chr, -start, -end) %>% 
  mutate(patient = str_remove_all(patient, "'")) %>% 
  group_by(start) %>% 
  summarise(counts = mean(counts))

xlist = c(25209431,25215437,25225614,
          25227234,25245274,25250751)
xend.list = c(25209911,25215560,25225773,
          25227412,25245395,25250803)

ggplot(kras.panc.cov,
       aes(start, counts)) + #, color = patient)) + 
  geom_line() +
  geom_segment(aes(x = 25209431,
                   xend = 25209911, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') + 
   geom_segment(aes(x = 25215437,
                   xend = 25215560, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') +
   geom_segment(aes(x = 25225614,
                   xend = 25225773, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') +
     geom_segment(aes(x = 25227234,
                   xend = 25227412, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') +
     geom_segment(aes(x = 25245274,
                   xend = 25245395, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') +
       geom_segment(aes(x = 25250751,
                   xend = 25250803, 
                   y = -0.5, yend = -0.5,
                   size = 3),
                   color = 'blue') +
  scale_color_npg()



tot <- 
  read_tsv('data/ftl.w.exons.aln', col_names = F) %>% 
  mutate(X7 = X2 + X7) %>% 
  select(position = 'X7', 
         cov = 'X8')

ggplot(tot,
      aes(position, cov), fill = 'grey') + 
  #geom_density(stat = 'identity', alpha = 0.3, fill = 'grey') +
  geom_bar(stat = 'identity', size = 5) 
  #scale_color_viridis_d(direction = -1) +
  #scale_fill_viridis_d(alpha = 0.3, direction = -1) + 
  scale_y_continuous(limits = c(-max(tot$cov),max(tot$cov))) +
  geom_hline(yintercept = 0, linetype = 'dotted')
```


## lncRNA analysis
### Filter data
```{r}
gencode.tx.norm.lncrna.counts <- 
  salmon.tx.normalized.counts %>% 
  merge(lnc.rna.names,
        by = 'tx') %>% 
  filter(gene %in% group.1.te.de.seq$gene)

gencode.mappings <- read_tsv('data/clean.map.txt',
                             col_names = F)

lnc.gencode.mappings <-
  gencode.mappings %>% 
  filter(X3 %in% gencode.tx.norm.lncrna.counts$tx)

te.gencode.mappings <- 
  read_tsv('data/te.clean.mappings.txt', 
           col_names = F)

lnc.te.gencode.mappings <- 
  te.gencode.mappings[,c(1,2,3)] %>% 
  merge(lnc.gencode.mappings,
        by = 'X1')

nrow(filter(lnc.gencode.mappings, X12 == 'NH:i:16'))
```
### Alu vs. Age
```{r}
te.salmon.tx.normalized.counts %>% 
  filter(grepl('Alu', tx)) %>% 
  gather(patient, count, -tx) %>% 
  merge(patient.info.encoded,
        by.x = 'patient',
        by.y = 'patient.id') %>% 
  group_by(age) %>% 
  filter(count > 0) %>% 
  summarize(avg.count = mean(count),
            sd.count = sd(count)) -> te.tx.alu.age.counts

salmon.real.tx.norm %>% 
  rownames_to_column('tx') %>% 
  gather(patient, count, -tx) %>% 
  merge(patient.info.encoded,
        by.x = 'patient',
        by.y = 'patient.id') %>% 
  group_by(age) %>%  
  mutate(count = as.numeric(count)) %>% 
  filter(count > 0) %>% 
  #select(age, count) %>% 
  summarize(avg.count = mean(count),
            sd.count = sd(count)) -> lnc.age.counts

lnc.age.counts %>% 
  ggplot(aes(as.factor(age), avg.count)) + 
  geom_col()

lnc.age.counts %>% 
  ggplot(aes(ymin = 0,
             ymax = avg.count + sd.count,
             x = as.factor(age), 
             avg.count)) + 
  geom_col() + 
  geom_errorbar()

te.tx.alu.age.counts %>% 
  ggplot(aes(as.factor(age), avg.count)) + 
  geom_col()

merge(te.tx.alu.age.counts,
      lnc.age.counts,
      by = 'age') %>% 
  ggplot(aes(avg.count.x, avg.count.y, color = as.factor(age))) + 
  geom_point()
  

```

## TCGA/TOIL
```{r}
toil.phenotypes <- 
  read_tsv('../KRAS.AALE/rna.seq.analysis/data/TcgaTargetGTEX_phenotype.txt') %>% 
  select(sample, 
         category = 'detailed_category',
         disease = 'primary disease or tissue',
         tissue = '_primary_site',
         type = '_sample_type',
         gender = '_gender',
         study = '_study') %>% 
  filter(study == 'TCGA') %>% 
  filter(tissue %in% c('Brain', 'Colon', 'Liver', 'White blood cell',
                       'Lung', 'Kidney', 'Pancreas', 'Soft tissue,Bone'))

table(toil.phenotypes$tissue)

toil.norm.counts <- 
  read_tsv('../KRAS.AALE/rna.seq.analysis/data/TCGA-GTEx-TARGET-gene-exp-counts.deseq2-normalized.log2')


toil.norm.counts.tidy <- 
  toil.norm.counts %>% 
  rename(gene = 'X1') %>% 
  gather(patient, count, -gene)

toil.norm.counts.phenotypes <- 
  toil.norm.counts.tidy %>% 
  filter(patient %in% toil.phenotypes$sample)

toil.norm.counts.phenotypes.merge <- 
  toil.norm.counts.phenotypes %>% 
  merge(toil.phenotypes,
        by.x = 'patient',
        by.y = 'sample') %>% 
  distinct()

```


## TE Overlap
```{r}
#### Generate total gencode reference
# need a reference with everything in it so I can calculate biotype proportions
whole.gene.gtf <- read_tsv('../KRAS.AALE/rna.seq.analysis/data/whole.genome.gtf',
                           col_names = c('chr', 'origin', 'type', 
                                         'start', 'stop', 'zero', 
                                         'strand', 'dot', 'gene.and.tx'))

# parse the gtf into more of a minimal .bed 
whole.gene.gtf.parsed <- 
  whole.gene.gtf[, c(1,3,4,5,7,9)] %>% 
  separate(gene.and.tx, ';', into = c('gene', 'tx')) %>% 
  separate(gene, '"', into = c('g', 'gene', 'q')) %>% 
  select(-g, -q) %>% 
  separate(tx, '"', into = c('t', 'tx', 'q')) %>% 
  select(-t, -q, -gene) 

# generate a annotated bed reference
salmon.normalized.counts %>% 
  gather(patient, norm.counts, -gene) %>% 
  merge(gencode.reference,
        by.x = 'gene',
        by.y = 'gene.name') %>% 
  filter(gene %in% lnc.rna.names$gene) %>% 
  merge(whole.gene.gtf.parsed,
        by = 'tx') -> salmon.normalized.ref.bed

# must use the start/stop from the gtf to get proper locations
salmon.normalized.ref.bed %>% 
  select(chr = 'chr.x', start = 'start.y', stop = 'stop.y', 
         tx.name, type, strand = 'strand.x') -> simple.salmon.normalized.ref.bed

write_tsv(simple.salmon.normalized.ref.bed, 'data/simple.salmon.normalized.ref.bed', 
          col_names = F)

bed.command <- 'cd data/ && bedtools intersect -a simple.salmon.normalized.ref.bed -b all.ucsc.rmsk.genes.bed -wao -s > reference.overlap.bed'
system(bed.command)
```

```{r}

## look at average counts TE vs LNC
lnc.te.summary.merge <- 
  salmon.tx.normalized.counts.tidy.lnc.summary %>% 
  spread(biotype, mean.counts) %>% 
  left_join(te.salmon.tx.norm.family.means) %>% 
  mutate(condition = ifelse(grepl('panc', patient.id), 'panc', 'ctrl'))

summary(lm(lncRNA~Alu, lnc.te.summary.merge))

ggplot(lnc.te.summary.merge,
       aes(lncRNA, Alu, color = condition)) + 
  geom_point() + 
  #stat_smooth(inherit.aes = F, 
  #            aes(lncRNA, Alu), 
  #            se = F, method = 'lm', size = 1, color = 'black') +
  scale_color_manual(name = '', values = c('#37b578ff',
                                           '#440154ff'))

ggsave(file.path(figure.dir.2,"te.and.lnc.compare.pdf"), 
             width = 2.5, height = 2.5, units = 'in')



overlap.col.names <- c('chr', 'tx.start', 'tx.end', 'tx', 'tx.type', 
                       'strand', 'chr.2', 'te.start', 'te.end',
                       'te', 'te.len', 'te.strand', 'te.tx.ol')

te.ref <- 
  read_csv('../KRAS.AALE/rna.seq.analysis/data/te.family.clade.csv', 
           col_names = c('te', 'family', 'clade')) %>% 
  distinct()
  
overlap.import.fxn <- function(ol.file, ref){
  overlap.output <- 
    read_tsv(ol.file, col_names = overlap.col.names) %>% 
    select(-chr.2, -te.strand) %>% 
    filter(!grepl('-rich', te)) %>% 
    filter(!grepl(')n', te))
  return(overlap.output)
}

# create a new counts df that is summarized to condition
salmon.normalized.ref.bed %>% 
  mutate(condition = ifelse(grepl('panc', patient), 'panc','ctrl')) %>% 
  group_by(condition, tx.name, biotype, tx.len) %>% 
  summarize(norm.counts.mean = mean(norm.counts),
            norm.counts.sd = sd(norm.counts)) -> by.condition.salmon.ref

# combine overlap with biotypes
all.rna.overlap <- 
  overlap.import.fxn('data/reference.overlap.bed', 
                     salmon.tx.normalized.counts.tidy) %>% 
  mutate(overlap = ifelse(te.tx.ol > 0, T, F)) %>% 
  mutate(exon.len = tx.end - tx.start) %>%
  merge(te.ref,
        by = 'te',all = T) %>% 
  mutate(prop.ol = ifelse(overlap == 0, 0, te.tx.ol/exon.len)) %>% 
  merge(by.condition.salmon.ref, 
        by.x = 'tx',
        by.y = 'tx.name') 
  
# compare overlap proportions across biotypes
ggplot(filter(all.rna.overlap, family %in% c(te.families.of.interest, NA)),
       aes(biotype, prop.ol, fill = family)) + 
  geom_boxplot()

table(salmon.normalized.ref.bed$biotype)


all.rna.overlap %>% 
  group_by(tx, condition)  %>% 
  filter(norm.counts.mean > 1) %>% 
  distinct() %>% 
  mutate(tx.prop.ol = sum(te.tx.ol)/tx.len) %>% 
  ungroup() %>% 
  select(tx, condition, norm.counts.mean, tx.prop.ol, biotype, family) -> test.ol #%>% 
  filter(biotype %in% c('lncRNA', 'protein_coding')) %>% 
  ggplot(aes(condition, norm.counts.mean, color = tx.prop.ol)) + 
  geom_sina(alpha = 0.6) + 
  scale_color_viridis_c() + 
  facet_wrap(~biotype, nrow = 1) + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1))



```

### TE Corr and PCA
```{r}
# prepare for matrix-style input
# filter out genes with low coefficient of variance and 
# genes that are present on chrY,M and Hemoglobin 

te.salmon.tx.norm <- 
  te.salmon.normalized.counts %>% 
  filter(! gene %in% chrM.txs$gene.name) %>% 
  column_to_rownames('gene')

te.salmon.tx.norm.family.means <- 
  te.salmon.normalized.counts %>% 
  gather(patient.id, counts, -gene) %>% 
  merge(te.ref,
        by.x = 'gene',
        by.y = 'te') %>% 
  group_by(patient.id, family) %>% 
  summarize(mean.counts = mean(counts)) %>% 
  spread(family, mean.counts)

te.ctrl.consistent.genes <- 
  as.data.frame(t(te.salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  filter(grepl('ctrl', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>%   
  #filter(gene %in% te.ref$te) %>% 
  #filter(grepl('Alu', gene)) %>% 
  #filter(gene %in% lnc.genes$gene.name) %>% 
  filter_at(vars(-gene), all_vars(. > 1)) %>% 
  #subset(rowMeans(.[,-c(1)]) > 13) %>% 
  mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>% 
  #filter(gene.cv > quantile(gene.cv, 0.70, na.rm = T)) %>%
  filter(gene.cv < quantile(gene.cv, 0.2, na.rm = T)) %>%
  select(-gene.cv)

te.panc.consistent.genes <- 
  as.data.frame(t(te.salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  filter(grepl('panc', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  #filter(gene %in% te.ref$te) %>% 
  #filter(grepl('Alu', gene)) %>% 
  filter_at(vars(-gene), all_vars(. > 1)) %>% 
  #subset(rowMeans(.[,-c(1)]) > 13) %>% 
  mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>% 
  filter(gene.cv > quantile(gene.cv, 0.70, na.rm = T),
         ! gene %in% ctrl.consistent.genes$gene) %>%
  #filter(gene.cv < quantile(gene.cv, 0.2, na.rm = T),
  #       ! gene %in% ctrl.consistent.genes$gene) %>%
  select(-gene.cv)

# filter to DE genes
te.de.genes.list <-  
  te.grouped.de.seq %>% 
  filter(abs(log2FoldChange) > 0)

# transpose data frame and add metadata while applying filters of choice
te.salmon.tx.norm.t <- 
  as.data.frame(t(te.salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  #filter(grepl('panc', patient.id)) %>% 
  filter(patient.id %in% biological.rep.patient.list) %>% 
  column_to_rownames('patient.id') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  filter_at(vars(-gene), any_vars(. > 0)) %>% 
  subset(rowMeans(.[,-c(1)]) > 13) %>% 
  filter(gene %in% te.de.genes.list$gene) %>% 
  #filter(gene %in% c(te.panc.consistent.genes$gene,
  #                   te.ctrl.consistent.genes$gene)) %>% 
  remove_rownames() %>% 
  column_to_rownames('gene') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('patient.id') %>% 
  merge(patient.info.encoded,
        by = 'patient.id') %>% 
  #left_join(patient.meds)  %>% 
  column_to_rownames('patient.id') 
```
### TE Correlation Matrices
```{r}
# plot correlation matrix sample vs. sample
ggcorrplot(cor(t(te.salmon.tx.norm.t)),
           hc.order = T, lab = F) + 
  scale_fill_viridis_c(direction = -1,
                       limits = c(-0.15, 1), option = 'viridis',
                       name = 'Pearson') + 
  ylab('') + 
  xlab('') +
            theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(0.4, 'cm'),
          panel.background = element_rect(fill = 'white'),
                axis.text.x = element_text(family = 'Helvetica', size = 6),
                axis.text.y = element_text(family = 'Helvetica', size = 6),
                axis.title.x = element_text(family = 'Helvetica', size = 6),
                axis.title.x.bottom = element_text(family = 'Helvetica', size = 6),
                axis.title.y = element_text(family = 'Helvetica', size = 6),
                legend.text = element_text(family = 'Helvetica', size = 6),
                legend.title = element_text(family = 'Helvetica', size = 6),
                axis.line = element_blank(),
                strip.text = element_text(family = 'Helvetica', 
                                          size = 6),
                strip.background = element_blank())

#ggsave("corr.pdf", width = 2.5, height = 1.5, units = 'in')
```

```{r}
tx.condition.mapping.volume.pca <-  
  te.salmon.tx.norm.t %>% 
  prcomp(scale=F)

#tx.condition.mapping.volume.pca <-  
#  patient.meds %>%
#  column_to_rownames('patient.id') %>% 
#  prcomp(scale=F)

#tx.condition.mapping.volume.pca <- 
#  lnc.te.summary.merge %>% 
#  column_to_rownames('patient.id') %>% 
#  select_if(is.numeric) %>% 
#  prcomp()


summary(tx.condition.mapping.volume.pca)
pcs.of.interest <- apply(tx.condition.mapping.volume.pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[1]
 
# convert output to dataframe
pca.out <- as.data.frame(tx.condition.mapping.volume.pca$x)
pca.out <- 
  pca.out %>% 
  rownames_to_column('patient.id') %>% 
  merge(patient.info.encoded, by = 'patient.id') %>%
  distinct() %>% 
  merge(salmon.tx.normalized.counts.tidy[,c(1,19)], by = 'patient.id') %>% 
  merge(te.salmon.tx.norm.family.means, by = 'patient.id') %>% 
  distinct() %>% 
  mutate(condition = ifelse(condition == 0 , 'ctrl', 'panc')) %>% 
  column_to_rownames('patient.id')

# summarize PC contributions
pca.out.summary <- 
  as.data.frame(summary(tx.condition.mapping.volume.pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)
# summarize PC contributions
pca.out.contributions <- 
  as.data.frame(tx.condition.mapping.volume.pca$rotation) %>% 
  rownames_to_column('tx') %>% 
  gather(pc, contrib, -tx) %>% 
  #merge(gencode.reference,
  #      by.x = 'tx',
  #      by.y = 'gene.name') %>% 
  group_by(pc) %>% 
  mutate(contrib = abs(contrib)) %>% 
  filter(contrib > quantile(contrib, 0.95))

# plot PC1 and PC2
ggplot(pca.out,
       aes(PC1, PC2, 
           color = Alu,
           label = paste(paste(condition,patient, sep = '.'),
                         paste0(age, 'yrs'),
                         sep = ', '))) + 
  geom_point(alpha = 0.75) + 
  scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                          digits = 3), sep = ' ')) +
 scale_color_viridis_c(direction = -1, name = 'Alu avg') + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  ggrepel::geom_text_repel(size = 1.5,
                            point.padding = 0.2,
                            box.padding = 0.11,
                            color = 'black',
                           segment.colour = NA) + 
  theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(0.4, 'cm'))

ggsave(file.path(figure.dir.2,"te.pca.pdf"), 
             width = 2.5, height = 2.5, units = 'in')

#ggsave("lnc.pca.plot.2.5x1.5.legend.pdf", width = 2.5, height = 1.5, units = 'in')

# print top contibutors per PC
pca.out.contributions %>%
  group_by(pc) %>%
  filter(pc %in% c("PC1", "PC2")) %>% 
  select(pc, contrib, tx) %>% 
  mutate(contrib = abs(contrib)) %>% 
  top_n(10, wt = contrib) %>% 
  arrange(pc, contrib)
```

#### TE Kmeans
```{r}
set.seed(234)

hc.pca <- hclust(as.dist(pca.out), method = "ward.D2")

pca.for.clustering <- 
  pca.out %>% 
  select(PC1, PC2)

elbow.plt(pca.for.clustering)

kmeans(pca.for.clustering, 3) -> 
  pca.k.means

pca.for.clustering %>% 
  rownames_to_column('patient.name') %>% 
  mutate(cluster = factor(pca.k.means$cluster)) %>%  
  merge(patient.info.encoded,
        by.x = 'patient.name',
        by.y = 'patient.id') %>%  
  merge(salmon.tx.normalized.counts.tidy[,c(1,19)], 
        by.x = 'patient.name',
        by.y = 'patient.id') %>% 
  distinct() %>% 
  mutate(condition = ifelse(condition == 0 , 'ctrl', 'panc')) ->
  pca.for.clustering

pca.for.clustering %>% 
  ggplot(aes(PC1, 
             PC2, 
             color = cluster, 
             label = paste(paste(condition,patient, sep = '.'),
                           #paste0(input_volume,'mL'),
                           paste0(age,'yrs'),
                           sep = ', '))) + 
  geom_point() + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                          digits = 3), sep = ' ')) +
  scale_color_viridis_d() + 
  ggrepel::geom_text_repel(size = 1.5,
                           point.padding = 0.1,
                          box.padding = 0.2,
                          segment.color = NA,
                          color = 'black') +
  
  theme(legend.position = 'right')


panc.salmon.tx.norm.t <- 
  salmon.tx.norm.t %>% 
  rownames_to_column('patient') %>% 
  filter(!grepl('ctrl', patient)) %>% 
  column_to_rownames('patient')
  
hc <- hclust(as.dist(salmon.tx.norm.t), method = "ward.D2")

plot(as.dendrogram(hc), 
     type = 'rectangle', 
     ylab = 'Height')

hc.dendro <- dendro_data(hc.pca, type = 'rectangle')
ggplot() +
  geom_segment(data=segment(hc.dendro), 
               aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_text(data=label(hc.dendro), 
            aes(x=x, y=y, label=label, hjust=-0.15), size=3) +
  coord_flip() + 
  ylab('Height') + 
  scale_y_reverse(expand=c(0.2, 0)) + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

ggsave(file.path(figure.dir.2,"te.pca.hclust.pdf"), 
             width = 3.5, height = 2.5, units = 'in')

plot(as.dendrogram(hc.pca), 
     type = 'rectangle', 
     ylab = 'Height')
```


#### TE NMF
```{r}
te.salmon.tx.norm %>% 
  rownames_to_column('tx') %>% 
  gather(patient, norm.counts, -tx) %>% 
  filter(grepl('panc', patient)) %>% 
  group_by(tx) %>% 
  summarise(sd = sd(norm.counts)) %>% 
  drop_na() %>% 
  filter(sd >= 0.8) -> sd.te.tx.gencode.panc

te.salmon.tx.norm.for.nmf <- 
  as.data.frame(t(te.salmon.tx.norm)) %>% 
  rownames_to_column('patient.id') %>% 
  filter(grepl('panc', patient.id)) %>% 
  merge(patient.info.encoded, by = 'patient.id') %>% 
  select(-condition) %>% 
  column_to_rownames('patient.id')

#te.salmon.tx.norm.for.nmf <- 
#te.salmon.tx.norm.for.nmf[, 
#                       colnames(te.salmon.tx.norm.for.nmf) %in% sd.te.tx.gencode.panc$tx]

te.salmon.tx.norm.for.nmf <- 
  te.salmon.tx.norm.for.nmf %>% 
  t() %>% 
  as.data.frame()
  

for(n in seq(2,10)){
  set.seed(123)
  te.nmf.out <- nmf(te.salmon.tx.norm.for.nmf, n, nrun = 10)
  fit(te.nmf.out)
  print(paste(n, cophcor(te.nmf.out), sep = ':'))
}

te.nmf.out <- nmf(te.salmon.tx.norm.for.nmf,2, nrun = 10)

featureScore(te.nmf.out)[extractFeatures(te.nmf.out)[[1]]]
featureScore(te.nmf.out)[extractFeatures(te.nmf.out)[[2]]]
featureScore(te.nmf.out)[extractFeatures(te.nmf.out)[[3]]]
featureScore(te.nmf.out)[extractFeatures(te.nmf.out)[[4]]]
featureScore(te.nmf.out)[extractFeatures(te.nmf.out)[[5]]]

qqnorm(featureScore(te.nmf.out))
qqline(featureScore(te.nmf.out))
summary(te.nmf.out)
basismap(te.nmf.out)
coefmap(te.nmf.out)
consensusmap(te.nmf.out)
```


### TE Group-based DEseq
```{r}
te.group.1.de.seq <- 
  read_csv('data/10.6.7.5.te.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 3')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')
  
te.group.2.de.seq <- 
  read_csv('data/8.4.3.te.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 2')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')

te.group.3.de.seq <- 
  read_csv('data/9.2.1.te.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.01) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() %>% 
  mutate(group = 'Cluster 4')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')

te.group.age.de.seq <- 
  read_csv('data/te.multi.var.panc.vs.ctrl.de.seq.csv') %>% 
  filter(padj < 0.1) %>% 
  rename(gene = 'X1') %>% 
  filter(! gene %in% c(hb.txs$gene.name, 
                       chr.Y.genes$gene.name, 
                       chrM.txs$gene.name),
         ! grepl(')n', gene), 
         ! grepl('-rich', gene)) %>% 
  distinct() #%>% 
  mutate(group = 'Cluster 4')#%>% 
  #merge(collison.subtype.genes,
  #      by.x = 'gene',
  #      by.y = 'Genes')

```

## TE DE Viz
```{r}
te.de.seq.plot <- function(de.data){
  de.data %>% 
    merge(te.ref,
          by.x = 'gene',
          by.y = 'te',
          all = T) %>% 
    distinct() %>% 
    filter(family %in% c(NA, te.families.of.interest)) %>% 
    mutate(family = ifelse(is.na(family), 'GENCODE', family)) %>% 
  ggplot(aes(log2FoldChange, 
             -log10(padj), 
             color = family)) +
  geom_point(alpha = 0.75, 
             size = 1) + 
  geom_rug(alpha = 0.2) +
  geom_vline(xintercept = 0, 
             color = 'black', 
             linetype = 'dotted') +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') + 
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', 
             alpha = 0.4, 
             size = 0.2) +
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) +
  geom_vline(xintercept = c(-1,1), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) + 
    scale_color_viridis_d() + 
    theme(legend.position = 'right',
          legend.key.width = unit(0.125,'cm'),
          legend.key.height = unit(1, 'cm')) #+ 
#    geom_text_repel(aes(label = ifelse(log2FoldChange > 2 & -log10(padj) > 3 & family == 'Alu', 
#                                   gene, '')),
#                    segment.color = NA,
#                    point.padding = NA) 
}

te.grouped.de.seq <- bind_rows(te.group.1.de.seq, 
                            te.group.2.de.seq,
                            te.group.3.de.seq)

te.de.seq.plot(te.group.age.de.seq)
ggsave(file.path(figure.dir.2,"all.groups.te.de.families.labels.pdf"), 
             width = 15, height = 15, units = 'in')

te.de.seq.plot <- function(de.data){
  de.data %>%
  mutate(biotype = ifelse(gene %in% te.ref$te, 'TE', 'GENCODE')) %>% 
  ggplot(aes(log2FoldChange, 
             -log10(padj), 
             color = biotype)) +
  geom_point(alpha = 0.35, Œ
             size = 1) + 
  geom_rug(alpha = 0.2) +
  geom_vline(xintercept = 0, 
             color = 'black', 
             linetype = 'dotted') +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') + 
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', 
             alpha = 0.4, 
             size = 0.2) +
  geom_hline(yintercept = -log10(0.05), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) +
  geom_vline(xintercept = c(-1,1), 
             linetype='dotted', alpha = 0.4, 
             size = 0.2) + 
  scale_color_manual(name = '', values = c('#37b578ff',
                                           '#440154ff')) + 
    scale_x_continuous(limits = c(-20,20)) + 
    theme(legend.position = 'bottom',
          legend.key.width = unit(1,'cm'),
          legend.key.height = unit(0.125, 'cm'))
}

te.de.seq.plot(te.group.1.de.seq)
ggsave(file.path(figure.dir.2,"group.1.te.de.biotype.pdf"), 
             width = 1.5, height = 2.5, units = 'in')
te.de.seq.plot(te.group.2.de.seq)
ggsave(file.path(figure.dir.2,"group.2.te.de.biotype.pdf"), 
             width = 1.5, height = 2.5, units = 'in')
te.de.seq.plot(te.group.3.de.seq)
ggsave(file.path(figure.dir.2,"group.3.te.de.biotype.pdf"), 
             width = 1.5, height = 2.5, units = 'in')


```

```{r}
panc.10.salmon.unmapped <- 
  read.table('clean.unmap.reads.txt')
panc.10.star.bed <- 
  read_tsv('panc.10.2.5.sorted.bed', col_names = F) %>% 
  separate(X4, sep = '/', into = c('read', 'mate')) %>% 
  filter(read %in% panc.10.salmon.unmapped$V1)


```

